modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/ags_solver.cc

```diff
@@ -75,6 +96,24 @@ AGSSolver::Solve()
     lbs_solver_.WriteRestartData();
 }
 
+double
+AGSSolver::ComputePhiErrorNorm() const
+{
+  double norm = 0.0;
+  auto& phi_new = lbs_solver_.PhiNewLocal();
+  for (int i = 0; i < phi_new.size(); ++i)
+  {
+    double val = phi_new[i] - phi_old_[i];
+    norm += val * val;
+  }
+
+  double global_norm = 0.0;
+  mpi_comm.all_reduce<double>(norm, global_norm, mpi::op::sum<double>());
+  global_norm = std::sqrt(global_norm);
+
+  return global_norm;
+}
+
 double
 AGSSolver::ComputePointwisePhiChange() const
```

Comment:
Isn't this implemented in `LBSSolver` in your Classic Richardson PR? Will there be duplicate routines when both PRs are in? Is that okay? Where should this routine live?

Reply:
Not quite. This isn't a test between lbs_solver::phi_new_local and lbs_solver::phi_old_local.

Reply:
Okay I am confused by a number of things now... In another comment you said we should be memory conscious and avoid making a copy of phi, but the AGS solver now stores a copy of phi. Where is it appropriate to make copies then? As far as I can tell, the difference between this routine and the one that is introduced in the Classic Richardson PR is that the Classic Richardson version in LBSSolver operates on group sets. I assume you by numerous global synchronizations you mean having to do that for every groupset, but couldn't that routine just be rewritten to handle a single groupset or all groupsets by doing something like passing the groupset ID and if it is -1 the operation is done on all groupsets?

Reply:
> Okay I am confused by a number of things now... In another comment you said we should be memory conscious and avoid making a copy of phi, but the AGS solver now stores a copy of phi. Where is it appropriate to make copies then? As far as I can tell, the difference between this routine and the one that is introduced in the Classic Richardson PR is that the Classic Richardson version in LBSSolver operates on group sets. I assume you by numerous global synchronizations you mean having to do that for every groupset, but couldn't that routine just be rewritten to handle a single groupset or all groupsets by doing something like passing the groupset ID and if it is -1 the operation is done on all groupsets?

The AGS solver has to have it's own `phi_old` that spans all groupsets. I don't know any other way to make it work. This was previously a PETSc vector (x_old) and is now an `std::vector` member of the AGS solver. In fact, the AGS solver is quite a bit leaner memory-wise since we previously had both `x` (phi) and `x_old` (phi_old) PETSc vectors (at least... I can't remember if there were others). In this case, the `phi_old` vector is necessary to implement the algorithm correctly. In the case of computing the L2 error, I don't need to make another copy of phi to compute that quantity and, IMHO, memory is at a premium, so it's better to duplicate 10 lines of code than to waste memory to use an existing routine.  I do agree that it's better to move the L2 routine that takes two vectors out of `ags_solver` and into math.h/math.cc. I didn't do this initially since I didn't expect it to get much use, but that is the better solution.

On the topic of Classic Richardson and pointwise computations, I could add functionality to the pointwise routine to handle all groupsets if you pass -1 as the id.  This seems a bit clunky to me, and would probably require duplicating a huge chunk of that routine to prevent having `if` statements in code that needs to be lean and fast and/or to avoid lots of global syncs. I'll look at it and see what can be done.

Reply:
`phi_new_local_` and `phi_old_local_` in `LBSSolver` span all groupsets. In the classic Richardson PR implementation this is why it is indexed `gsi + g`. You are already pulling in `lbs_solver_.PhiNewLocal()`, why can you not use `lbs_solver_.PhiOldLocal()`?

On the pointwise routine, you are already duplicating the code, it just depends on where you do it. Alternatively, you could have an override that does not take a groupset in `LBSSolver`. A second option would be to loop over groupsets at the outermost level and do the check there. A third option would be to create a vector of groups outside at the top of the function based on the function input. There is already an if statement in the innermost portion of the code.

Reply:
I fully understand the argument of allowing for some duplicate code to allow for better performance. The issues I am trying to raise here are that 1) we should try to generalize where we can first, and 2) if we do duplicate code, we should think carefully about where it lives.

In this case, the differences between the two routines are 1) access to a member variable, 2) groupset vs. all groups, and 3) living in `LBSSolver` which does not have iterative schemes implemented vs. living in something in `iterative_methods`. On 1), we should not duplicate code when using a function argument will accomplish the same task. On 2) and 3) If the performance hit of an if statement in the outermost loop or defining group bounds at the top of the function is too large and a duplicate routine is necessary, we should think more carefully about where each routine lives. IMHO, both should live in LBSSolver or both should live in a class defined within `iterative_methods`.

Reply:
> `phi_new_local_` and `phi_old_local_` in `LBSSolver` span all groupsets. In the classic Richardson PR implementation this is why it is indexed `gsi + g`. You are already pulling in `lbs_solver_.PhiNewLocal()`, why can you not use `lbs_solver_.PhiOldLocal()`?

Yes, I understand that `phi_new_local` and `phi_old_local` span all groupsets. But, using `phi_old_local_` won't work. It does not contain the value of phi from the previous AGS iteration.

> On the pointwise routine, you are already duplicating the code, it just depends on where you do it. Alternatively, to not duplicate code, you could have an override that does not take a groupset in `LBSSolver`. A second option would be to loop over groupsets at the outermost level and do the check there. A third option would be to create a vector of groups outside at the top of the function based on the function input.

Yes, I think I like option 3 the best. I'll take a look... like I said, I'm sure it can be fixed, I'm just not a fan of passing -1 as a groupset id.

Reply:
>On 1), we should not duplicate code when using a function argument will accomplish the same task. 

I agree, and I'll move this routine into `math.h`/`math.cc`.

>On 2) and 3) If the performance hit of an if statement in the outermost loop or defining group bounds at the top of >the function is too large and a duplicate routine is necessary, we should think more carefully about where each >routine lives. IMHO, both should live in `LBSSolver` or both should live in an appropriate class defined within >`iterative_methods`.

Yes. As I mentioned in a comment in my Classic Richardson PR, I already have a new PR ready to go that pulls out all of the io functionality of `LBSSolver` into separate routines. It would be nice to do something similar with things like convergence checking... and hopefully it works out to be as clean as the io stuff did. But, I am actively trying not to add any additional code to `LBSSolver` unless it absolutely needs to live there.

Reply:
> It does not contain the value of phi from the previous AGS iteration.

Ah, this is a good point I did not think about. Each WGS solver updates the respective portion of the phi vectors. Once we get around to addressing where solutions should live, these types of things should be much more clear.

> Yes, I think I like option 3 the best. I'll take a look... like I said, I'm sure it can be fixed, I'm just not a fan of passing -1 as a groupset id.

I'm not a huge fan either. If you look at `ComputeLeakage` in `DiscreteOrdinatesSolver` you could do something similar. Pass a vector of groupset IDs. If it is empty, you compute for all groupsets, if it is not, you use the groupset IDs from the argument.

Reply:
I believe I've resolved this issue. I've moved the convergence checks out of LBSSolver and into separate routines in `lbs_solver/iterative_algorithms/convergence.h/.cc`.